Unit11 course project

ChepaKN opened the pull request.

ChepaKN committed:
Реализовал изменение вопроса(с ответами). (удаление старого из базы и сохранение нового).

ChepaKN committed:
При редактировании вопроса, если вопрос с таким id почему то не найден добавил читабельное исключение.

ChepaKN committed:
В процессе поддержки сессий

ChepaKN committed:
Добавлена поддержка сессий с передачей клиенту списка вопросов с вариантами ответов.
Далее нужно приделать валидацию ответов и сохранение выбранных вариантов ответов в базу.
С этим есть трудности, т.к. клиент почему то 2 раза запрашивает вопросы, изза этого после завершении сессии их не удается провалидировать, т.к. вопросы не соответствуют ответам.

ChepaKN committed:
Merge pull request #14 from SiberianIntegrationSystems/master

Client bug fixed

ChepaKN committed:
Добавил рандомизацию порядка ответов для новой сесии + признак isCorrect() для каждого ответа устанавливается в false;

ChepaKN committed:
Добавил валидацию сессии. Осталось сохранить выбранные ответы в базу и написать тесты.

ChepaKN committed:
Добавил сохранение в базу id выбранного ответа с ссылками на id сессии и id ответа. Не уверен что такой формат удобен. я бы добавил что нибудь типа самого текста вопроса.

ChepaKN committed:
Испрален баг с редактированием вопроса. Подчищен неиспользуемый код

ChepaKN committed:
Исправил баг с обнуление флага правильности ответа при отправке новых вопросов для сессии. Раньше сброс этого влага попадал в базу.

ChepaKN committed:
Добавил тесты для сервисов, и интеграционный тест для приложения в целом. Осталось добавить тесты на rest контроллеры.

ChepaKN committed:
Добавил тест на сервис журнала

ChepaKN committed:
change gitignore

ChepaKN committed:
weather app gitignore changed

Mrak2017 commented on unit-11/app-server/src/main/java/com/github/siberianintegrationsystems/restApp/controller/SessionRestController.java:
```diff
@@ -0,0 +1,37 @@
+
+@RestController
+@CrossOrigin(origins = "*")
+@RequestMapping("api")
```
можно было указать "api/session", тогда бы не пришлось писать session для каждого метода

Mrak2017 commented on unit-11/app-server/src/main/java/com/github/siberianintegrationsystems/restApp/controller/dto/SessionItemDTO.java:
```diff
@@ -0,0 +1,23 @@
+public class SessionItemDTO extends JournalItemDTO {
+
+    public String   name;
+    public Date     insertDate;
```
лучше использовать LocalDate для даты или LocalDateTime если нужно ещё и время.

Mrak2017 commented on unit-11/app-server/src/main/java/com/github/siberianintegrationsystems/restApp/data/QuestionRepository.java:
```diff
@@ -9,4 +10,10 @@
+    int questionsPerSession = 2;
+    @Query(value = "SELECT * FROM QUESTION ORDER BY RAND() LIMIT " +
+            questionsPerSession, nativeQuery = true)
+    List<Question> findRandQuestions();
```
Можно использовать такой подход (findTop10ByLastname) https://docs.spring.io/spring-data/jpa/docs/current/reference/html/#repositories.limit-query-result

ChepaKN replied:
Здесь проблема скорее не в колличестве вопросов, а в случайном выборе их из базы. Про findTop я читал, но решил оставить так.

Mrak2017 commented on unit-11/app-server/src/main/java/com/github/siberianintegrationsystems/restApp/entity/SessionEvent.java:
```diff
@@ -0,0 +1,41 @@
+
+
+    @Column
+    private Date date;
```
аналогично, лучше использовать LocalDate или LocalDateTime

Mrak2017 commented on unit-11/app-server/src/main/java/com/github/siberianintegrationsystems/restApp/service/QuestionServiceImpl.java:
```diff
@@ -34,11 +37,68 @@ public QuestionsItemDTO createQuestion(QuestionsItemDTO dto) {
+                questionRepository.findRandQuestions()
+                    .stream()
+                    .map(question -> {
+                                //перемешаем ответы
```
Интересная задумка :)

Mrak2017 commented on unit-11/app-server/src/main/java/com/github/siberianintegrationsystems/restApp/service/QuestionServiceImpl.java:
```diff
@@ -34,11 +37,68 @@ public QuestionsItemDTO createQuestion(QuestionsItemDTO dto) {
+                    )
+                    .collect(Collectors.toList());
+
+        //Скроем правильные ответы от хитрых пользователей
```
Отлично!

Mrak2017 commented on unit-11/app-server/src/main/java/com/github/siberianintegrationsystems/restApp/service/SessionServiceImpl.java:
```diff
@@ -0,0 +1,122 @@
+        Question question = questionRepository
+                .findById(Long.parseLong(questionSessionDTO.id))
+                .orElseThrow(() ->
+                        new RuntimeException(String.format("Не найдем вопрос с id%s",
```
"Не найдеМ" или "Не найдеН"?

Mrak2017 commented on unit-11/app-server/src/main/java/com/github/siberianintegrationsystems/restApp/service/SessionServiceImpl.java:
```diff
@@ -0,0 +1,122 @@
+
+
+    private double validateOneQuestion(QuestionSessionDTO questionSessionDTO){
+        //Результат в диапазоне [0:1]
```
Комментарий не ясен. Раз он относится ко всему методу, надо было его разместить над методом

Mrak2017 commented on unit-11/app-server/src/main/java/com/github/siberianintegrationsystems/restApp/service/SessionServiceImpl.java:
```diff
@@ -0,0 +1,122 @@
+        if(m == 0) {
+            throw new RuntimeException(
+                    String.format("Не найден правильный ответ для вопроса с id: %d",
+                            question.getId()));
```
Интересная проверка, но я бы сказал, что она должна происходить на этапе сохранения вопроса с ответами

Mrak2017 commented on unit-11/app-server/src/main/java/com/github/siberianintegrationsystems/restApp/service/SessionServiceImpl.java:
```diff
@@ -0,0 +1,122 @@
+        }else if(n == m){
+            throw new RuntimeException(
+                    String.format("Все ответы на вопрос с id: %d верные",
+                            question.getId()));
```
1) А в чем тут проблема? Возможно составитель вопроса хотел именно так?
2) Опять же, проверку надо делать не на проверке сессии, а на сохранении вопроса

Mrak2017 commented on unit-11/app-server/src/main/java/com/github/siberianintegrationsystems/restApp/service/SessionServiceImpl.java:
```diff
@@ -0,0 +1,122 @@
+        //верно выбранных
+        double k = 0;
+        //неверно выбранных
+        double w = 0;
```
Переменные лучше называть более говорящими именами, тогда и комментарии с описанием бы не понадобились

Mrak2017 commented on unit-11/app-server/src/main/java/com/github/siberianintegrationsystems/restApp/service/SessionServiceImpl.java:
```diff
@@ -0,0 +1,122 @@
+                    w++;
+                }
+            }
+            //todo: как быть с невыбранными верными ответами?
```
В каком кейсе формула их не учтет?

Mrak2017 commented on unit-11/app-server/src/main/java/com/github/siberianintegrationsystems/restApp/service/SessionServiceImpl.java:
```diff
@@ -0,0 +1,122 @@
+        //сохраним выбранные ответы
+        List<Answer> answers = new ArrayList<>();
+        for(QuestionSessionDTO q : sessionDTO.questionsList){
+            answers.addAll(q.answersList.stream()
```
Я бы результат обработки потока вынес в отдельную переменную, которую уже подавал в .addAll.
Для читабельности

Mrak2017 commented on unit-11/app-server/src/main/java/com/github/siberianintegrationsystems/restApp/service/SessionServiceImpl.java:
```diff
@@ -0,0 +1,122 @@
+        }
+
+        for (Answer a : answers){
+            selectedAnswerRepository.save(new SelectedAnswer(a, sessionEvent));
```
почему бы сразу не сохранять внутри прошлого for? тогда не понадобится промежуточная переменная со списком ответов.

Mrak2017 commented on unit-11/app-server/src/main/java/com/github/siberianintegrationsystems/restApp/service/SessionServiceImpl.java:
```diff
@@ -0,0 +1,122 @@
+        }
+
+        //Locale.US для того чтобы при преобразовании в строку разделитель был точкой а не запятой, иначе клиент ругается
+        return String.format(Locale.US, "%.2f", result);
```
Можно было бы сделать replace внутри самой DTO ответа.

Mrak2017 commented on unit-11/app-server/src/test/java/com/github/siberianintegrationsystems/restApp/RestAppApplicationTests.java:
```diff
@@ -1,16 +1,60 @@
 	@Test
-	public void contextLoads() {
+	public void contextLoads(){
+		//Проверим что все зависимости подтянулись
```
Были случаи когда не подтягивались? За это отвечает спринг, зачем его перепроверять?

Mrak2017 replied:
Тем более в дальнейших тестах этой проверки уже нет. Следуя логике, зависимости могли не подтянуться и там.

Mrak2017 commented on unit-11/app-server/src/test/java/com/github/siberianintegrationsystems/restApp/service/JournalServiceImplTest.java:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    @Test
+    public void getJournal() {
```
лучше называть метод testGetJournal, иначе создается впечатление, что это вспомогательный метод, который должен вернуть журнал.

Mrak2017 commented on unit-11/app-server/src/test/java/com/github/siberianintegrationsystems/restApp/service/QuestionServiceImplTest.java:
```diff
@@ -0,0 +1,131 @@
+        questionsItemDTO.name = "Сколько ног у Паука";
+        questionsItemDTO.answers = Arrays.asList(answerItemDTO1, answerItemDTO2);
+
+        QuestionsItemDTO createdQuestion = questionService.createQuestion(questionsItemDTO);
```
разве в методе init не был создан точно такой же вопрос?

Mrak2017 commented on unit-11/app-server/src/test/java/com/github/siberianintegrationsystems/restApp/service/QuestionServiceImplTest.java:
```diff
@@ -0,0 +1,131 @@
+
+        int answersLenBeforeEdit = createdQuestion.answers.size();
+        createdQuestion.name = createdQuestion.name + " some text";
+        createdQuestion.answers.remove(0);
```
можно было бы ещё проверить, что не удастся сохранить вопрос, без ответов. Не критично

Mrak2017 commented on unit-11/app-server/src/test/java/com/github/siberianintegrationsystems/restApp/service/QuestionServiceImplTest.java:
```diff
@@ -0,0 +1,131 @@
+        List<QuestionsItemDTO> questionsItemDTOList = questionService.getQuestionsForSession();
+        assertNotNull(questionsItemDTOList);
+
+        //все ответы к вопросу должны быть false;
```
я бы уточнил в комментарии, что ситуация штатная, т.к. мы заранее очищаем правильность ответов, перед отправкой на клиент. Иначе поведение, что нет ни одного правильного ответа по началу вводит в заблуждение

Mrak2017 reviewed (changes requested):
В целом работа хорошая, посмотрите замечания.
---
Михаил Павлов

ChepaKN commented:
Здравствуйте, Михаил. Пробежался по замечаниям, практичен со всем
согласен..:) есть необходимость присылать исправления?

ср, 29 апр. 2020 г., 12:25 Mikhail <notifications@github.com>:

> *@Mrak2017* requested changes on this pull request.
> В целом работа хорошая, посмотрите замечания.
>
> Михаил Павлов
> ------------------------------
>
> In
> unit-11/app-server/src/main/java/com/github/siberianintegrationsystems/restApp/controller/SessionRestController.java
> <https://github.com/ChepaKN/JavaSIS-3.20/pull/19#discussion_r417051725>:
>
> > @@ -0,0 +1,37 @@
> +package com.github.siberianintegrationsystems.restApp.controller;
> +
> +
> +import com.github.siberianintegrationsystems.restApp.controller.dto.QuestionsItemDTO;
> +import com.github.siberianintegrationsystems.restApp.controller.dto.session.SessionDTO;
> +import com.github.siberianintegrationsystems.restApp.service.QuestionService;
> +import com.github.siberianintegrationsystems.restApp.service.SessionService;
> +import org.springframework.web.bind.annotation.*;
> +import java.util.List;
> +
> +@RestController
> +@CrossOrigin(origins = "*")
> +@RequestMapping("api")
>
> можно было указать "api/session", тогда бы не пришлось писать session для
> каждого метода
> ------------------------------
>
> In
> unit-11/app-server/src/main/java/com/github/siberianintegrationsystems/restApp/controller/dto/SessionItemDTO.java
> <https://github.com/ChepaKN/JavaSIS-3.20/pull/19#discussion_r417056908>:
>
> > @@ -0,0 +1,23 @@
> +package com.github.siberianintegrationsystems.restApp.controller.dto;
> +
> +import com.github.siberianintegrationsystems.restApp.controller.dto.journal.JournalItemDTO;
> +import com.github.siberianintegrationsystems.restApp.entity.SessionEvent;
> +import java.util.Date;
> +
> +public class SessionItemDTO extends JournalItemDTO {
> +
> +    public String   name;
> +    public Date     insertDate;
>
> лучше использовать LocalDate для даты или LocalDateTime если нужно ещё и
> время.
> ------------------------------
>
> In
> unit-11/app-server/src/main/java/com/github/siberianintegrationsystems/restApp/data/QuestionRepository.java
> <https://github.com/ChepaKN/JavaSIS-3.20/pull/19#discussion_r417058321>:
>
> > @@ -9,4 +10,10 @@
>          extends CrudRepository<Question, Long> {
>
>      List<Question> findByNameContainingIgnoreCase(String search);
> +
> +    //todo: как бы вынести эту константу в настройки
> +    int questionsPerSession = 2;
> +    @Query(value = "SELECT * FROM QUESTION ORDER BY RAND() LIMIT " +
> +            questionsPerSession, nativeQuery = true)
> +    List<Question> findRandQuestions();
>
> Можно использовать такой подход (findTop10ByLastname)
> https://docs.spring.io/spring-data/jpa/docs/current/reference/html/#repositories.limit-query-result
> ------------------------------
>
> In
> unit-11/app-server/src/main/java/com/github/siberianintegrationsystems/restApp/entity/SessionEvent.java
> <https://github.com/ChepaKN/JavaSIS-3.20/pull/19#discussion_r417058540>:
>
> > @@ -0,0 +1,41 @@
> +package com.github.siberianintegrationsystems.restApp.entity;
> +
> +import javax.persistence.*;
> +import java.time.LocalDateTime;
> +import java.util.Date;
> +
> +@Entity
> +public class SessionEvent extends BaseEntity {
> +
> +
> +    @Column
> +    private Date date;
>
> аналогично, лучше использовать LocalDate или LocalDateTime
> ------------------------------
>
> In
> unit-11/app-server/src/main/java/com/github/siberianintegrationsystems/restApp/service/QuestionServiceImpl.java
> <https://github.com/ChepaKN/JavaSIS-3.20/pull/19#discussion_r417061370>:
>
> >              answerRepository.save(answer);
>          }
>
>          return new QuestionsItemDTO(question,
>                  answerRepository.findByQuestion(question));
>      }
> +
> +    @Override
> +    public List<QuestionsItemDTO> getQuestionsForSession(){
> +
> +
> +        List<QuestionsItemDTO> questionsItemDTOList =
> +                questionRepository.findRandQuestions()
> +                    .stream()
> +                    .map(question -> {
> +                                //перемешаем ответы
>
> Интересная задумка :)
> ------------------------------
>
> In
> unit-11/app-server/src/main/java/com/github/siberianintegrationsystems/restApp/service/QuestionServiceImpl.java
> <https://github.com/ChepaKN/JavaSIS-3.20/pull/19#discussion_r417061782>:
>
> > +
> +        List<QuestionsItemDTO> questionsItemDTOList =
> +                questionRepository.findRandQuestions()
> +                    .stream()
> +                    .map(question -> {
> +                                //перемешаем ответы
> +                                // todo: (как то не очень, переделать бы...)
> +                                List<Answer> answers = answerRepository.findByQuestion(question);
> +                                Collections.shuffle(answers);
> +                                return new QuestionsItemDTO(question,
> +                                        answers);
> +                            }
> +                    )
> +                    .collect(Collectors.toList());
> +
> +        //Скроем правильные ответы от хитрых пользователей
>
> Отлично!
> ------------------------------
>
> In
> unit-11/app-server/src/main/java/com/github/siberianintegrationsystems/restApp/service/SessionServiceImpl.java
> <https://github.com/ChepaKN/JavaSIS-3.20/pull/19#discussion_r417062006>:
>
> > +                              AnswerRepository answerRepository,
> +                              SelectedAnswerRepository selectedAnswerRepository) {
> +        this.sessionEventRepository = sessionEventRepository;
> +        this.questionRepository = questionRepository;
> +        this.answerRepository = answerRepository;
> +        this.selectedAnswerRepository = selectedAnswerRepository;
> +    }
> +
> +
> +
> +    private double validateOneQuestion(QuestionSessionDTO questionSessionDTO){
> +        //Результат в диапазоне [0:1]
> +        Question question = questionRepository
> +                .findById(Long.parseLong(questionSessionDTO.id))
> +                .orElseThrow(() ->
> +                        new RuntimeException(String.format("Не найдем вопрос с id%s",
>
> "Не найдеМ" или "Не найдеН"?
> ------------------------------
>
> In
> unit-11/app-server/src/main/java/com/github/siberianintegrationsystems/restApp/service/SessionServiceImpl.java
> <https://github.com/ChepaKN/JavaSIS-3.20/pull/19#discussion_r417062225>:
>
> > +
> +
> +    public SessionServiceImpl(SessionEventRepository sessionEventRepository,
> +                              QuestionRepository questionRepository,
> +                              AnswerRepository answerRepository,
> +                              SelectedAnswerRepository selectedAnswerRepository) {
> +        this.sessionEventRepository = sessionEventRepository;
> +        this.questionRepository = questionRepository;
> +        this.answerRepository = answerRepository;
> +        this.selectedAnswerRepository = selectedAnswerRepository;
> +    }
> +
> +
> +
> +    private double validateOneQuestion(QuestionSessionDTO questionSessionDTO){
> +        //Результат в диапазоне [0:1]
>
> Комментарий не ясен. Раз он относится ко всему методу, надо было его
> разместить над методом
> ------------------------------
>
> In
> unit-11/app-server/src/main/java/com/github/siberianintegrationsystems/restApp/service/SessionServiceImpl.java
> <https://github.com/ChepaKN/JavaSIS-3.20/pull/19#discussion_r417062513>:
>
> > +                .orElseThrow(() ->
> +                        new RuntimeException(String.format("Не найдем вопрос с id%s",
> +                                questionSessionDTO.id)));
> +
> +        //всего ответов
> +        double n = questionSessionDTO.answersList.size();
> +        //всего верных для этого вопроса
> +        double m = answerRepository.findByQuestion(question)
> +                .stream()
> +                .filter(Answer::getCorrect)
> +                .count();
> +
> +        if(m == 0) {
> +            throw new RuntimeException(
> +                    String.format("Не найден правильный ответ для вопроса с id: %d",
> +                            question.getId()));
>
> Интересная проверка, но я бы сказал, что она должна происходить на этапе
> сохранения вопроса с ответами
> ------------------------------
>
> In
> unit-11/app-server/src/main/java/com/github/siberianintegrationsystems/restApp/service/SessionServiceImpl.java
> <https://github.com/ChepaKN/JavaSIS-3.20/pull/19#discussion_r417062759>:
>
> > +        //всего ответов
> +        double n = questionSessionDTO.answersList.size();
> +        //всего верных для этого вопроса
> +        double m = answerRepository.findByQuestion(question)
> +                .stream()
> +                .filter(Answer::getCorrect)
> +                .count();
> +
> +        if(m == 0) {
> +            throw new RuntimeException(
> +                    String.format("Не найден правильный ответ для вопроса с id: %d",
> +                            question.getId()));
> +        }else if(n == m){
> +            throw new RuntimeException(
> +                    String.format("Все ответы на вопрос с id: %d верные",
> +                            question.getId()));
>
>
>    1. А в чем тут проблема? Возможно составитель вопроса хотел именно так?
>    2. Опять же, проверку надо делать не на проверке сессии, а на
>    сохранении вопроса
>
> ------------------------------
>
> In
> unit-11/app-server/src/main/java/com/github/siberianintegrationsystems/restApp/service/SessionServiceImpl.java
> <https://github.com/ChepaKN/JavaSIS-3.20/pull/19#discussion_r417065785>:
>
> > +                .count();
> +
> +        if(m == 0) {
> +            throw new RuntimeException(
> +                    String.format("Не найден правильный ответ для вопроса с id: %d",
> +                            question.getId()));
> +        }else if(n == m){
> +            throw new RuntimeException(
> +                    String.format("Все ответы на вопрос с id: %d верные",
> +                            question.getId()));
> +        }
> +
> +        //верно выбранных
> +        double k = 0;
> +        //неверно выбранных
> +        double w = 0;
>
> Переменные лучше называть более говорящими именами, тогда и комментарии с
> описанием бы не понадобились
> ------------------------------
>
> In
> unit-11/app-server/src/main/java/com/github/siberianintegrationsystems/restApp/service/SessionServiceImpl.java
> <https://github.com/ChepaKN/JavaSIS-3.20/pull/19#discussion_r417067742>:
>
> > +        double w = 0;
> +        List<AnswerSessionDTO> selectedAnswers = questionSessionDTO.answersList;
> +
> +        for(AnswerSessionDTO selectedAns : selectedAnswers){
> +            Answer savedAns = answerRepository.findById(Long.parseLong(selectedAns.id))
> +                    .orElseThrow(() -> new RuntimeException(
> +                            String.format("Не найден ответ с id: %s", selectedAns.id)));
> +
> +            if(selectedAns.isSelected){
> +                if(savedAns.getCorrect()){
> +                    k++;
> +                }else{
> +                    w++;
> +                }
> +            }
> +            //todo: как быть с невыбранными верными ответами?
>
> В каком кейсе формула их не учтет?
> ------------------------------
>
> In
> unit-11/app-server/src/main/java/com/github/siberianintegrationsystems/restApp/service/SessionServiceImpl.java
> <https://github.com/ChepaKN/JavaSIS-3.20/pull/19#discussion_r417068451>:
>
> > +    @Override
> +    public String validateSession(SessionDTO sessionDTO){
> +
> +        Double result = sessionDTO.questionsList.stream()
> +                .map(this::validateOneQuestion).reduce(0.0, Double::sum);
> +
> +        result = result / sessionDTO.questionsList.size() * 100.0;
> +
> +        //сохраним сессию
> +        SessionEvent sessionEvent = new SessionEvent(sessionDTO.name, result);
> +        sessionEventRepository.save(sessionEvent);
> +
> +        //сохраним выбранные ответы
> +        List<Answer> answers = new ArrayList<>();
> +        for(QuestionSessionDTO q : sessionDTO.questionsList){
> +            answers.addAll(q.answersList.stream()
>
> Я бы результат обработки потока вынес в отдельную переменную, которую уже
> подавал в .addAll.
> Для читабельности
> ------------------------------
>
> In
> unit-11/app-server/src/main/java/com/github/siberianintegrationsystems/restApp/service/SessionServiceImpl.java
> <https://github.com/ChepaKN/JavaSIS-3.20/pull/19#discussion_r417069027>:
>
> > +        //сохраним сессию
> +        SessionEvent sessionEvent = new SessionEvent(sessionDTO.name, result);
> +        sessionEventRepository.save(sessionEvent);
> +
> +        //сохраним выбранные ответы
> +        List<Answer> answers = new ArrayList<>();
> +        for(QuestionSessionDTO q : sessionDTO.questionsList){
> +            answers.addAll(q.answersList.stream()
> +                    .filter(answerSessionDTO -> answerSessionDTO.isSelected)
> +                    .map(answerSessionDTO -> answerRepository.findById(Long.parseLong(answerSessionDTO.id))
> +                            .orElseThrow(RuntimeException::new))
> +                    .collect(Collectors.toList()));
> +        }
> +
> +        for (Answer a : answers){
> +            selectedAnswerRepository.save(new SelectedAnswer(a, sessionEvent));
>
> почему бы сразу не сохранять внутри прошлого for? тогда не понадобится
> промежуточная переменная со списком ответов.
> ------------------------------
>
> In
> unit-11/app-server/src/main/java/com/github/siberianintegrationsystems/restApp/service/SessionServiceImpl.java
> <https://github.com/ChepaKN/JavaSIS-3.20/pull/19#discussion_r417069240>:
>
> > +        //сохраним выбранные ответы
> +        List<Answer> answers = new ArrayList<>();
> +        for(QuestionSessionDTO q : sessionDTO.questionsList){
> +            answers.addAll(q.answersList.stream()
> +                    .filter(answerSessionDTO -> answerSessionDTO.isSelected)
> +                    .map(answerSessionDTO -> answerRepository.findById(Long.parseLong(answerSessionDTO.id))
> +                            .orElseThrow(RuntimeException::new))
> +                    .collect(Collectors.toList()));
> +        }
> +
> +        for (Answer a : answers){
> +            selectedAnswerRepository.save(new SelectedAnswer(a, sessionEvent));
> +        }
> +
> +        //Locale.US для того чтобы при преобразовании в строку разделитель был точкой а не запятой, иначе клиент ругается
> +        return String.format(Locale.US, "%.2f", result);
>
> Можно было бы сделать replace внутри самой DTO ответа.
> ------------------------------
>
> In
> unit-11/app-server/src/test/java/com/github/siberianintegrationsystems/restApp/RestAppApplicationTests.java
> <https://github.com/ChepaKN/JavaSIS-3.20/pull/19#discussion_r417069389>:
>
> >
>  	@Test
> -	public void contextLoads() {
> +	public void contextLoads(){
> +		//Проверим что все зависимости подтянулись
>
> Были случаи когда не подтягивались? За это отвечает спринг, зачем его
> перепроверять?
> ------------------------------
>
> In
> unit-11/app-server/src/test/java/com/github/siberianintegrationsystems/restApp/RestAppApplicationTests.java
> <https://github.com/ChepaKN/JavaSIS-3.20/pull/19#discussion_r417069538>:
>
> >
>  	@Test
> -	public void contextLoads() {
> +	public void contextLoads(){
> +		//Проверим что все зависимости подтянулись
>
> Тем более в дальнейших тестах этой проверки уже нет. Следуя логике,
> зависимости могли не подтянуться и там.
> ------------------------------
>
> In
> unit-11/app-server/src/test/java/com/github/siberianintegrationsystems/restApp/service/JournalServiceImplTest.java
> <https://github.com/ChepaKN/JavaSIS-3.20/pull/19#discussion_r417070051>:
>
> > +    private JournalService journalService;
> +
> +    private final String journalID = JournalServiceImpl.QUESTIONS_JOURNAL_ID;
> +    private final String journalName = "Незнайка и Java";
> +
> +    @Before
> +    public void init(){
> +        Journal journal = new Journal();
> +        journal.setId(journalID);
> +        journal.setName(journalName);
> +        journal.setDefaultPageSize(15L);
> +        journalRepository.save(journal);
> +    }
> +
> +    @Test
> +    public void getJournal() {
>
> лучше называть метод testGetJournal, иначе создается впечатление, что это
> вспомогательный метод, который должен вернуть журнал.
> ------------------------------
>
> In
> unit-11/app-server/src/test/java/com/github/siberianintegrationsystems/restApp/service/QuestionServiceImplTest.java
> <https://github.com/ChepaKN/JavaSIS-3.20/pull/19#discussion_r417070664>:
>
> > +        answer1.setCorrect(true);
> +        answerRepository.save(answer1);
> +
> +        Answer answer2 = new Answer();
> +        answer2.setName("16");
> +        answer2.setCorrect(false);
> +        answerRepository.save(answer2);
> +
> +        AnswerItemDTO answerItemDTO1 = new AnswerItemDTO(answer1);
> +        AnswerItemDTO answerItemDTO2 = new AnswerItemDTO(answer2);
> +
> +        QuestionsItemDTO questionsItemDTO = new QuestionsItemDTO();
> +        questionsItemDTO.name = "Сколько ног у Паука";
> +        questionsItemDTO.answers = Arrays.asList(answerItemDTO1, answerItemDTO2);
> +
> +        QuestionsItemDTO createdQuestion = questionService.createQuestion(questionsItemDTO);
>
> разве в методе init не был создан точно такой же вопрос?
> ------------------------------
>
> In
> unit-11/app-server/src/test/java/com/github/siberianintegrationsystems/restApp/service/QuestionServiceImplTest.java
> <https://github.com/ChepaKN/JavaSIS-3.20/pull/19#discussion_r417070922>:
>
> > +        AnswerItemDTO answerItemDTO2 = new AnswerItemDTO(answer2);
> +
> +        QuestionsItemDTO questionsItemDTO = new QuestionsItemDTO();
> +        questionsItemDTO.name = "Сколько ног у Паука";
> +        questionsItemDTO.answers = Arrays.asList(answerItemDTO1, answerItemDTO2);
> +
> +        QuestionsItemDTO createdQuestion = questionService.createQuestion(questionsItemDTO);
> +
> +        //заодно проверим работу с репозиториями
> +        assertTrue(answerRepository.existsById(answer1.getId()));
> +        assertTrue(answerRepository.existsById(answer2.getId()));
> +        assertTrue(questionRepository.existsById(Long.parseLong(createdQuestion.id)));
> +
> +        int answersLenBeforeEdit = createdQuestion.answers.size();
> +        createdQuestion.name = createdQuestion.name + " some text";
> +        createdQuestion.answers.remove(0);
>
> можно было бы ещё проверить, что не удастся сохранить вопрос, без ответов.
> Не критично
> ------------------------------
>
> In
> unit-11/app-server/src/test/java/com/github/siberianintegrationsystems/restApp/service/QuestionServiceImplTest.java
> <https://github.com/ChepaKN/JavaSIS-3.20/pull/19#discussion_r417071359>:
>
> > +        createdQuestion.answers.remove(0);
> +        QuestionsItemDTO editQuestion = questionService.editQuestion(createdQuestion);
> +
> +        assertTrue(editQuestion.name.contains(" some text"));
> +        assertEquals(answersLenBeforeEdit - 1,
> +                editQuestion.answers.size());
> +
> +    }
> +
> +
> +    @Test
> +    public void getQuestionsForSession() {
> +        List<QuestionsItemDTO> questionsItemDTOList = questionService.getQuestionsForSession();
> +        assertNotNull(questionsItemDTOList);
> +
> +        //все ответы к вопросу должны быть false;
>
> я бы уточнил в комментарии, что ситуация штатная, т.к. мы заранее очищаем
> правильность ответов, перед отправкой на клиент. Иначе поведение, что нет
> ни одного правильного ответа по началу вводит в заблуждение
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ChepaKN/JavaSIS-3.20/pull/19#pullrequestreview-402348852>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKWSJO6TE64JOGVDFXJNXC3RO623NANCNFSM4MPYVXRQ>
> .
>

Mrak2017 commented:
Да, желательно исправить, чтобы довести работу до логического завершения

ChepaKN committed:
Исправления:
1. Date в LocalDateTame
2. Проверка наличия правильного ответа(ов) при создании/редактировании вопроса
3. Читабельные названия переменных при валидации вопроса

ChepaKN committed:
Исправления:
1. Добавил генерацию исключения при создания вопроса без ответов
2. Проверка на то что указан правильный ответ делается при сохранении вопроса
3. Добавил тест на сохранение вопроса без ответов

ChepaKN committed:
Подчистил. При создании вопроса без ответов проверяется спиков ответов к вопросу проверяеся на null и isEmpty

ChepaKN committed:
Исправления:
1. Добавил генерацию исключения, если вдруг с  клиента прилетит сессия без вопросов.
2. Удалил лишние комментарии

ChepaKN commented:
https://github.com/ChepaKN/JavaSIS-3.20/pull/20

Здравствуйте, высылаю исправления по замечаниям

ср, 29 апр. 2020 г., 12:52 Mikhail <notifications@github.com>:

> Да, желательно исправить, чтобы довести работу до логического завершения
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ChepaKN/JavaSIS-3.20/pull/19#issuecomment-621004544>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKWSJO423HCQ3EHZZTUA7PLRO66BXANCNFSM4MPYVXRQ>
> .
>

ChepaKN committed:
Merge pull request #20 from ChepaKN/unit11_revision

Unit11 revision
